system/core adaptations for Boot2Gecko.

diff --git a/include/private/android_filesystem_config.h b/include/private/android_filesystem_config.h
index a058686..b2de29b 100755
--- a/include/private/android_filesystem_config.h
+++ b/include/private/android_filesystem_config.h
@@ -178,6 +178,9 @@ static struct fs_path_config android_dirs[] = {
 ** and will allow partial matches.
 */
 static struct fs_path_config android_files[] = {
+    { 00775, AID_ROOT,      AID_ROOT,      "system/b2g/b2g" },
+    { 00775, AID_ROOT,      AID_ROOT,      "system/b2g/updater" },
+    { 00775, AID_ROOT,      AID_ROOT,      "system/b2g/plugin-container" },
     { 00440, AID_ROOT,      AID_SHELL,     "system/etc/init.goldfish.rc" },
     { 00550, AID_ROOT,      AID_SHELL,     "system/etc/init.goldfish.sh" },
     { 00550, AID_ROOT  ,    AID_SYSTEM,    "system/etc/init.qcom.sdio.sh" },
diff --git a/rootdir/init.rc b/rootdir/init.rc
index 811ba36..612fa22 100755
--- a/rootdir/init.rc
+++ b/rootdir/init.rc
@@ -6,6 +6,7 @@ on early-init
 
 # create mountpoints
     mkdir /mnt 0775 root system
+    import /init.b2g.rc
 
 on init
 
@@ -15,7 +16,7 @@ loglevel 3
 
 # setup the global environment
     export PATH /sbin:/vendor/bin:/system/sbin:/system/bin:/system/xbin
-    export LD_LIBRARY_PATH /vendor/lib:/system/lib
+    symlink /dev/socket/rild /dev/socket/rilproxy
     export ANDROID_BOOTLOGO 1
     export ANDROID_ROOT /system
     export ANDROID_ASSETS /system/app
-- 
1.7.6

